sos/report/plugins/shmcli.py

```diff
@@ -0,0 +1,112 @@
+# This file is part of the sos project: https://github.com/sosreport/sos
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# version 2 of the GNU General Public License.
+#
+# See the LICENSE file in the source distribution for further information.
+
+from sos.report.plugins import Plugin, IndependentPlugin, PluginOpt
+
+
+class SHMcli(Plugin, IndependentPlugin):
+    """shmcli pulls hardware information from PowerVault/Dell
+    Storage JBOD's attached to server.
+    It provides information of the adapters, emms, drives,
+    enclosures, fans, power supplies and the sensory data of
+    temp, voltage, and current sensors.
+    """
+
+    short_desc = 'Dell Server Hardware Manager'
+
+    plugin_name = 'shmcli'
+    profiles = ('system', 'storage', 'hardware',)
+
+    files = ('/opt/dell/ServerHardwareManager/' +
+             'ServerHardwareManagerCLI/bin/shmcli',)
+    option_list = [
+        PluginOpt('debug', default=False, desc='capture support debug data')
+    ]
+
+    def setup(self):
+        cmd = "/opt/dell/ServerHardwareManager/" \
+              "ServerHardwareManagerCLI/bin/shmcli"
+
+        subcmds = [
+            'list adapters',
+            'list physical enclosures',
+            'list failed drives'
+        ]
+
+        for subcmd in subcmds:
+            self.add_cmd_output(
+                "%s %s" % (cmd, subcmd),
+                suggest_filename="shmcli_%s" % (subcmd))
+
+        models = []
+
+        # Get the storage hardware models
+        result = self.exec_cmd('lsscsi -g')
+        if result['status'] == 0:
+            for line in result['output'].splitlines():
+                if (line.split()[2] == 'DELL'):
+                    models.append(line.split()[3])
+        models = list(set(models))
+
+        subcmds = [
+            'list emms',
+            'list drawers',
+            'list emm slots',
+            'list drive slots',
+            'list fans',
+            'list temp sensors',
+            'list voltage sensors',
+            'list current sensors',
+            'list power supplies',
+            'info enclosure'
+        ]
+
+        if self.get_option("debug"):
+            subcmds.append("getdebugcli")
+
+        result = self.collect_cmd_output(
+                    '%s list enclosures' % (cmd),
+                    suggest_filename='shmcli_list_enclosures')
+        if result['status'] == 0:
+            for line in result['output'].splitlines():
+                for model in models:
+                    if (model in line.split()):
+                        adapt_index = line.split()[-1]
+                        enc_index = line.split()[0]
```

Comment:
Can you pastebin sample output of this command? I'd like to see if we can tighten this up a little bit.

Reply:
```
nb-appliance:/log # /opt/dell/ServerHardwareManager/ServerHardwareManagerCLI/bin/shmcli list enclosures

list physical enclosures - Executing command..
Encl#            Name     Service Tag Encl Overall Sts              WWN   # Adapters Connected
-----            ----     ----------- ----------------              ---   --------------------
    0          MD1400         JSWVQN3               OK 520474729ae6ab00                      1

list physical enclosures - Command execution complete.
```

Reply:
```python3
if result['status'] == 0:
    for line in result['output'].splitlines()[2:-2]:
        _line = line.split()
        if any(m in _line for m in models):
            adapt_index = _line[-1]
            enc_index = _line[0]
            [...]
```